irohad/ordering/impl/on_demand_ordering_service_impl.hpp

```diff
@@ -8,23 +8,100 @@
 
 #include "ordering/on_demand_ordering_service.hpp"
 
+#include <queue>
+#include <shared_mutex>
+#include <unordered_map>
+
+#include <tbb/concurrent_queue.h>
+
+#include "logger/logger.hpp"
+
 namespace iroha {
   namespace ordering {
     class OnDemandOrderingServiceImpl : public OnDemandOrderingService {
      public:
-      // --------------------- | OnDemandOrderingService | ---------------------
+      OnDemandOrderingServiceImpl(
+          size_t transaction_limit,
+          size_t number_of_proposals = 3,
+          const transport::RoundType &initial_round = std::make_pair(1, 1));
+
+      // --------------------- | OnDemandOrderingService |_---------------------
 
       void onCollaborationOutcome(RoundOutput outcome,
                                   transport::RoundType round) override;
 
       // ----------------------- | OdOsNotification | --------------------------
 
-      void onTransactions(
-          std::vector<std::shared_ptr<shared_model::interface::Transaction>>
-              transactions) override;
+      void onTransactions(const CollectionType &transactions) override;
+
+      boost::optional<ProposalType> onRequestProposal(
+          transport::RoundType round) override;
+
+     private:
+      /**
+       * Type of stored transactions
+       */
+      using TransactionType =
```

Comment:
Is it possible to use `unique_ptr` here? The object is created in transport, and then moved to a proposal. It seems to express the ownership more clearly.

Reply:
Maybe also use `CollectionType::value_type` to share the type with notification?

Reply:
tbb won't allow :(